File: tests/falling.test.mjs
import test from 'node:test';
import assert from 'node:assert/strict';
import fs from 'node:fs/promises';
import {getDefaults,normalizeOptions,SCHEMAS,PRESETS} from '../src/core/schema.js';
import {FALL_SHAPES} from '../src/core/fall-schema.js';
import {createFallSystem} from '../src/core/fall-system.js';
import {skeletonShapes,ringGeometry} from '../src/core/status-geometry.js';
import {resolveMotion} from '../src/core/motion.js';
import {EFFECT_NAMES} from '../src/core/catalog.js';

const options=(preset='snow',patch={})=>normalizeOptions('fall',patch,getDefaults('fall',preset));
const simulation=(o=options())=>{const system=createFallSystem(o);system.resize(800,400);system.step(0);return system;};
const clone=system=>structuredClone(system.particles);

test('Falling presets have valid material-specific shapes',()=>{
 assert.ok(EFFECT_NAMES.includes('fall'));
 for(const name in PRESETS.fall){const o=options(name);assert.ok(FALL_SHAPES[o.kind].includes(o.shape),name);}
 for(const kind of SCHEMAS.fall.kind.values)assert.ok(Object.values(PRESETS.fall).some(p=>(p.kind||'snow')===kind));
});
test('Changing material clears an incompatible particle shape',()=>{
 const o=normalizeOptions('fall',{kind:'leaves'},options('snowflakes'));assert.equal(o.shape,'mixed');
 assert.equal(normalizeOptions('fall',{shape:'maple'},o).shape,'maple');
});
test('Falling option budgets clamp finite values and preserve zero',()=>{
 const o=options('snow',{maxParticles:19.8,fallSpeed:0,density:0,size:NaN,gust:Infinity});
 assert.equal(o.maxParticles,20);assert.equal(o.fallSpeed,0);assert.equal(o.density,0);assert.equal(o.size,11);assert.ok(Number.isFinite(o.gust));
});
test('Seeded particles and trajectories are reproducible',()=>{
 const a=simulation(),b=simulation();a.step(4);b.step(4);assert.deepEqual(clone(a),clone(b));
 const c=simulation(options('snow',{seed:53}));assert.notDeepEqual(clone(a),clone(c));
});
test('Falling motion integrates equivalently at different frame rates',()=>{
 for(const preset of ['snow','autumn-leaves','cherry-petals','dandelion-seeds']){
  const a=simulation(options(preset)),b=simulation(options(preset));
  for(let i=1;i<=300;i++)a.step(i/60);for(let i=1;i<=50;i++)b.step(i/10);
  for(let i=0;i<a.particles.length;i++)for(const key of ['x','y','rotation'])assert.ok(Math.abs(a.particles[i][key]-b.particles[i][key])<1e-8,`${preset}/${key}`);
 }
});
test('Wind-only and still trajectories honor explicit zero fall speed',()=>{
 const a=simulation(options('snow',{fallSpeed:0,gust:0,sway:0,tumble:0,wind:20,depth:0}));const before=clone(a);a.step(1);
 a.particles.forEach((p,i)=>{assert.equal(p.y,before[i].y);assert.ok(Math.abs(p.x-before[i].x-20)<1e-10);assert.equal(p.rotation,before[i].rotation);});
});
test('Particle drift continues between samples without paused-time catch-up',()=>{
 const a=simulation();a.step(1);const before=clone(a);a.step(1);assert.deepEqual(clone(a),before);a.step(1.1);assert.notDeepEqual(clone(a),before);
});
test('Hard particle budgets and adaptive quality bound large viewports',()=>{
 const a=simulation(options('snow',{density:120,maxParticles:230}));a.resize(20000,10000);assert.equal(a.particles.length,230);
 a.resize(800,400);a.setQuality(.25);assert.equal(a.particles.length,96);
 a.setOptions(options('snow',{maxParticles:0}));assert.equal(a.particles.length,0);
});
test('Particle positions remain finite and outside recycling is bounded',()=>{
 const a=simulation(options('snow-flurry',{wind:-300,fallSpeed:500,sway:120,gust:200}));a.step(100000);
 assert.ok(a.getState().recycled>0);for(const p of a.particles){assert.ok(Number.isFinite(p.x)&&Number.isFinite(p.y));assert.ok(p.x>-200&&p.x<1000);assert.ok(p.y>-200&&p.y<600);}
});
test('Resizing preserves normalized placement and render-time changes do not reseed',()=>{
 const a=simulation(),p=a.particles[0],x=p.x,y=p.y;a.resize(1600,800);assert.equal(p.x,x*2);assert.equal(p.y,y*2);
 a.setOptions(options('snow',{wind:100,color:'#ff0000'}));assert.equal(a.particles[0],p);
});
test('Material changes and rewind reset deterministically, destroy clears arrays',()=>{
 const a=simulation();a.step(1);a.step(0);assert.deepEqual(clone(a),clone(simulation()));
 a.setOptions(options('autumn-leaves'));assert.ok(a.particles.every(p=>FALL_SHAPES.leaves.includes(p.shape)));
 a.destroy();assert.equal(a.particles.length,0);
});
test('Falling particles default to hiding decoration under reduced motion',()=>{
 for(const preset in PRESETS.fall){const o=options(preset);assert.equal(resolveMotion('fall',o,true),'hide');assert.equal(resolveMotion('fall',{...o,motion:'always'},true),'full');assert.equal(resolveMotion('fall',{...o,reducedMotion:'still'},true),'still');}
});
test('Profile bars share a column and uniform visible gaps',()=>{
 const o=getDefaults('status','avatar-placeholder'),shapes=skeletonShapes(700,320,o),[avatar,...bars]=shapes;
 assert.equal(bars.length,3);for(const b of bars)assert.ok(b.x>=avatar.x+avatar.w+o.avatarGap-1e-9);
 for(let i=1;i<bars.length;i++){assert.equal(bars[i].x,bars[0].x);assert.ok(Math.abs(bars[i].y-bars[i-1].y-bars[i-1].h-o.gap)<1e-9);}
 const top=Math.min(...shapes.map(s=>s.y)),bottom=Math.max(...shapes.map(s=>s.y+s.h));assert.equal((top+bottom)/2,160);
});
test('Profile sizing works with one to eight lines and narrow text columns',()=>{
 for(const width of [80,320,800])for(let lines=1;lines<=8;lines++){
  const shapes=skeletonShapes(width,350,{...getDefaults('status','avatar-placeholder'),lines});assert.equal(shapes.length,lines+1);
  for(const s of shapes)assert.ok(Object.values(s).every(Number.isFinite)&&s.w>0&&s.h>0);
 }
});
test('Progress and loading rings default to complete circular tracks',()=>{
 assert.equal(SCHEMAS.status.sweep.default,360);for(const preset of ['progress-ring','loading-ring'])assert.equal(getDefaults('status',preset).sweep,360);
 assert.equal(getDefaults('status','progress-ring').value,.72);assert.ok(getDefaults('status','loading-ring').indeterminate);
});
test('Ring sizing accounts for line thickness without cropping its circumference',()=>{
 for(const size of [10,100,400])for(const anchorX of [.05,.5,.95]){
  const o={...getDefaults('status','progress-ring'),size,anchorX},g=ringGeometry(180,90,o);
  assert.ok(g.radius>=0);assert.ok(g.radius+g.thickness/2<=Math.min(180*anchorX,180*(1-anchorX),45)+1e-9);
 }
});
test('Package and exported entry versions agree',async()=>{
 const pkg=JSON.parse(await fs.readFile(new URL('../package.json',import.meta.url),'utf8'));assert.match(pkg.version,/^\d+\.\d+\.\d+(?:-[\w.]+)?$/);
 for(const p of ['../src/index.js','../src/selective.js'])assert.ok((await fs.readFile(new URL(p,import.meta.url),'utf8')).includes(`version='${pkg.version}'`));
});

File: tests/falling_test.py
"""Native Canvas/layout regression tests for falling particles and status fixes.

Uses locally supplied pages in restricted environments; no WebGL stub is involved.
If a production website is built, its minified assets are used for the public demo.
"""
import base64
import json
import mimetypes
import os
from pathlib import Path
import re
import shutil
from urllib.parse import urlparse
from playwright.sync_api import sync_playwright

ROOT = Path(__file__).resolve().parents[1]
OUT = ROOT / 'tests/validation'
OUT.mkdir(exist_ok=True)
SITE = ROOT / 'build/site' if (ROOT / 'build/site/index.html').exists() else ROOT
checks = []


def check(name, condition):
    """Record a regression check and stop on failure."""
    checks.append({'name': name, 'passed': bool(condition)})
    assert condition, name


def install_demo(page):
    """Supply exact built JS/CSS bytes without a top-level network navigation."""
    text = (SITE / 'index.html').read_text()
    scripts = ['site/theme.js', 'dist/frontend-toolkit.js', 'site/logic.js', 'site/ui.js']
    for file in scripts:
        text = re.sub(r'<script\b[^>]*\bsrc=[\'"]' + re.escape(file) + r'[\'"][^>]*>\s*</script>', '', text)
    image = 'data:image/svg+xml;base64,' + base64.b64encode((SITE / 'assets/backdrop.svg').read_bytes()).decode()
    for file in ['site/style.css', 'site/design.css']:
        css = (SITE / file).read_text().replace('../assets/backdrop.svg', image)
        text = re.sub(r'<link\b[^>]*\bhref=[\'"]' + re.escape(file) + r'[\'"][^>]*>', lambda _: '<style>' + css + '</style>', text)
    text = text.replace('src="site/favicon.svg"', 'src="data:image/svg+xml;base64,' + base64.b64encode((SITE / 'site/favicon.svg').read_bytes()).decode() + '"')
    page.set_content(text)
    for file in scripts:
        page.add_script_tag(content=(SITE / file).read_text())
    page.wait_for_function('window.ToolkitDemo?.element?.controller')


def main():
    with sync_playwright() as pw:
        browser = pw.chromium.launch(headless=True, executable_path=os.environ.get('CHROMIUM_PATH') or shutil.which('chromium'))
        page = browser.new_page(viewport={'width': 1200, 'height': 900}, reduced_motion='no-preference')
        errors = []
        page.on('pageerror', lambda e: errors.append(str(e)))
        page.set_content('<style>body{margin:0;background:#112129}#mount{position:relative;width:760px;height:350px;border-radius:20px}</style><div id="mount"></div>')
        page.add_script_tag(content=(ROOT / 'dist/frontend-toolkit.js').read_text())
        page.evaluate('''()=>{
          const FT=FrontendToolkit;window.r=null;
          window.make=(effect,preset,patch={})=>{r?.destroy();mount.replaceChildren();r=FT['create'+effect[0].toUpperCase()+effect.slice(1)+'Renderer'](mount,{...FT.getDefaults(effect,preset),quality:1,dprCap:1,adaptive:false,...patch});r.resize(760,350);r.render(0);return r;};
          window.signature=canvas=>{const p=canvas.getContext('2d').getImageData(0,0,canvas.width,canvas.height).data;let hash=2166136261,ink=0;for(let i=0;i<p.length;i++){hash=Math.imul(hash^p[i],16777619);if(i%4===3&&p[i])ink++;}return {hash:hash>>>0,ink};};
        }''')
        for preset in page.evaluate('Object.keys(FrontendToolkit.PRESETS.fall)'):
            result = page.evaluate('''preset=>{make('fall',preset);const a=signature(r.canvas);r.render(1.2);const b=signature(r.canvas);return{a,b,state:r.getState()};}''', preset)
            check(preset + ': paints visible particles', result['a']['ink'] > 30)
            check(preset + ': motion changes actual pixels', result['a']['hash'] != result['b']['hash'])
            check(preset + ': reuses bounded sprite cache', result['state']['spriteBuilds'] == 1 and 1 <= result['state']['spriteCount'] <= 9)
            page.locator('#mount').screenshot(path=str(OUT / ('fall-' + preset + '.png')))
        results = page.evaluate('''()=>{
          const results=[],check=(name,ok)=>{results.push({name,passed:!!ok});if(!ok)throw Error(name);};
          make('fall','autumn-leaves');const a=signature(r.canvas),count=r.getState().spriteBuilds;r.render(0);check('Redrawing the same time is stable',signature(r.canvas).hash===a.hash);
          r.setOptions({wind:75,fallSpeed:90,density:30});r.render(1);check('Velocity and density do not rebuild sprites',r.getState().spriteBuilds===count);
          r.setOptions({color:'#49aed1'});r.render(1);check('Changing color rebuilds the sprite cache once',r.getState().spriteBuilds===count+1);
          r.setOptions({kind:'snow',shape:'flake',colorMode:'single'});r.render(1);check('Switching material and palette produces one matching sprite',r.getState().kind==='snow'&&r.getState().shape==='flake'&&r.getState().spriteCount===1);
          r.setOptions({fallSpeed:0,wind:0,gust:0,swaySpeed:0,tumble:0});r.render(2);const still=signature(r.canvas);r.render(3);check('Explicitly still configuration sleeps without changing pixels',!r.animated&&signature(r.canvas).hash===still.hash);
          r.setOptions({density:0});r.render(3);check('Zero-density overlay clears and sleeps',!r.animated&&r.getState().particleCount===0&&signature(r.canvas).ink===0);
          r.setOptions({density:120,maxParticles:47});r.resize(20000,8000);r.render(3);check('Large falling surfaces obey both pixel and particle caps',r.canvas.width<=2048&&r.canvas.height<=2048&&r.canvas.width*r.canvas.height<=1400000&&r.getState().particleCount<=47);
          const old=r.canvas;r.destroy();check('Destroy releases the canvas and sprite cache',!old.isConnected&&old.width===1&&r.getState().spriteCount===0&&r.getState().particleCount===0);
          for(const preset of ['progress-ring','loading-ring']){
            make('status',preset,{anchorX:.5,anchorY:.5,value:0,indeterminate:false});
            const o=FrontendToolkit.getDefaults('status',preset),ctx=r.canvas.getContext('2d'),radius=o.size/2-o.thickness/2;
            let closed=true;for(let deg=0;deg<360;deg+=5){const a=deg*Math.PI/180,x=Math.round(380+radius*Math.cos(a)),y=Math.round(175+radius*Math.sin(a));if(!ctx.getImageData(x-1,y-1,3,3).data.some((v,i)=>i%4===3&&v>100))closed=false;}
            check(preset+': complete 360-degree track, including former gap',closed&&r.getState().value===0);
          }
          make('status','progress-ring',{anchorX:.5,anchorY:.5});check('Full track does not falsify actual progress',r.getState().value===.72);
          make('status','loading-ring',{anchorX:.5,anchorY:.5});const before=signature(r.canvas);r.render(.6);check('Circular loading indicator retains animated highlight',r.animated&&signature(r.canvas).hash!==before.hash);
          return results;
        }''')
        checks.extend(results)
        page.evaluate("make('status','avatar-placeholder',{anchorX:.5,anchorY:.5,shimmer:false,width:.7})")
        page.locator('#mount').screenshot(path=str(OUT / 'profile-layout.png'))
        # Validate the actual profile pixels, not only the pure layout helper.
        boxes = page.evaluate('''()=>{
          const c=r.canvas,d=c.getContext('2d').getImageData(0,0,c.width,c.height).data,seen=new Uint8Array(c.width*c.height),result=[];
          for(let n=0;n<seen.length;n++){if(seen[n]||!d[n*4+3])continue;const queue=[n];seen[n]=1;let minX=c.width,maxX=0,minY=c.height,maxY=0;
            for(let i=0;i<queue.length;i++){const k=queue[i],x=k%c.width,y=Math.floor(k/c.width);minX=Math.min(minX,x);maxX=Math.max(maxX,x);minY=Math.min(minY,y);maxY=Math.max(maxY,y);
              for(const j of [x? k-1:-1,x<c.width-1?k+1:-1,y?k-c.width:-1,y<c.height-1?k+c.width:-1])if(j>=0&&!seen[j]&&d[j*4+3]){seen[j]=1;queue.push(j);}}
            result.push({x:minX,y:minY,w:maxX-minX+1,h:maxY-minY+1});}
          return result;
        }''')
        avatar = max(boxes, key=lambda b: b['h'])
        bars = sorted([b for b in boxes if b != avatar], key=lambda b: b['y'])
        check('Profile paints one avatar and three separate text bars', len(boxes) == 4 and len(bars) == 3)
        check('Every profile bar is right of the avatar with matching gaps', len({b['x'] for b in bars}) == 1 and bars[0]['x'] >= avatar['x'] + avatar['w'] + 9 and bars[1]['y'] - bars[0]['y'] - bars[0]['h'] == bars[2]['y'] - bars[1]['y'] - bars[1]['h'])
        page.close()
        page = browser.new_page(viewport={'width':1440,'height':1000},reduced_motion='reduce')
        page.on('pageerror', lambda e: errors.append(str(e)))
        install_demo(page)
        for width in [320,390,768,1440]:
            page.set_viewport_size({'width':width,'height':1000})
            for study in ['mesh','bokeh','sunbeams','grid','spotlight']:
                # IDs are stable catalog identifiers, not preset labels.
                identifier = page.evaluate('''name=>ToolkitSiteModel.catalog.find(i=>i.preset===name)?.id''', study)
                page.evaluate('(id)=>ToolkitDemo.select(id)',identifier)
                page.locator('#stage').scroll_into_view_if_needed()
                page.wait_for_timeout(60)
                layout = page.evaluate('''()=>{
                  const s=document.querySelector('#stage').getBoundingClientRect(),c=document.querySelector('.example-card').getBoundingClientRect();
                  return {cx:Math.abs(c.x+c.width/2-s.x-s.width/2),cy:Math.abs(c.y+c.height/2-s.y-s.height/2),inside:c.top>=s.top-1&&c.bottom<=s.bottom+1,overflow:document.documentElement.scrollWidth>innerWidth+1};
                }''')
                check(f'{study} at {width}px: centered, fully visible card', layout['cx'] < 1 and layout['cy'] < 1 and layout['inside'])
                check(f'{study} at {width}px: no page overflow', not layout['overflow'])
            if width in [390,1440]:
                page.evaluate("ToolkitDemo.select(ToolkitSiteModel.catalog.find(i=>i.preset==='bokeh').id)")
                page.locator('#stage').screenshot(path=str(OUT / f'centered-bokeh-{width}.png'))
        page.set_viewport_size({'width':1100,'height':950})
        for identifier in ['snowfall','autumn-leaves','falling-petals','drifting-seeds']:
            page.evaluate('(id)=>ToolkitDemo.select(id)',identifier)
            page.locator('#stage').scroll_into_view_if_needed()
            page.wait_for_function("ToolkitDemo.element.controller?.renderer?.getState().particleCount>0")
            check(identifier + ': public full motion despite reduced system', page.evaluate("ToolkitDemo.element.options.motion==='always'&&ToolkitDemo.element.getStats().animating"))
            check(identifier + ': only material-specific shapes offered', page.evaluate("Array.from(document.querySelector('#control-shape').options).every(o=>FrontendToolkit.SCHEMAS.fall.shape.valuesByKind[ToolkitDemo.element.options.kind].includes(o.value))"))
            check(identifier + ': native touch panning retained',page.evaluate("getComputedStyle(ToolkitDemo.element).touchAction==='auto'"))
            page.locator('#motion-reduced').click()
            page.wait_for_timeout(60)
            check(identifier + ': reduced mode hides decoration and stops rendering',page.evaluate("ToolkitDemo.element.getStats().motionPresentation==='hide'&&!ToolkitDemo.element.getStats().animating"))
            page.locator('#motion-full').click()
            page.wait_for_timeout(70)
            check(identifier + ': full mode can recreate released decoration',page.evaluate("ToolkitDemo.element.controller.renderer?.getState().particleCount>0"))
        # Inspect the advanced link generated by the public demo.
        page.evaluate("ToolkitDemo.select('snowfall')")
        check('Full settings link carries the new study and shape',page.evaluate("JSON.parse(decodeURIComponent(document.querySelector('#lab-link').getAttribute('href').split('#lab=')[1])).id==='snowfall'"))
        page.close()
        # Selective loading reads the generated module tree, without fetching GPU effects.
        page=browser.new_page(reduced_motion='no-preference')
        page.on('pageerror',lambda e:errors.append(str(e)))
        requested=[]
        def route_local(route):
            path=urlparse(route.request.url).path.lstrip('/')
            requested.append(path)
            file=(ROOT/path).resolve()
            if file.is_relative_to(ROOT) and file.is_file():
                route.fulfill(body=file.read_bytes(),headers={'Content-Type':mimetypes.guess_type(file)[0] or 'application/octet-stream','Access-Control-Allow-Origin':'*'})
            else: route.fulfill(status=404,body='Not found')
        page.route('https://toolkit.test/**',route_local)
        page.set_content('<style>ft-fall{display:block;width:600px;height:300px}</style><ft-fall id="falling" preset="snowflakes" loading="manual"><button>Native button</button></ft-fall><script type="module" src="https://toolkit.test/dist/frontend-toolkit.loader.js"></script>')
        page.wait_for_function("!!customElements.get('ft-fall')")
        check('Manual falling element does not fetch renderer early', not any(p.endswith('/effects/fall.js') for p in requested))
        page.evaluate("document.querySelector('ft-fall').load()")
        page.wait_for_function("document.querySelector('ft-fall').controller?.renderer?.getState().particleCount>0")
        check('Manual falling activation loads exactly one renderer module',sum(p.endswith('/effects/fall.js') for p in requested)==1)
        check('Selective falling activation does not download unrelated GPU renderers',not any('/effects/'+name+'.js' in p for p in requested for name in ['fire','water','aurora','sea','fluid']))
        page.emulate_media(reduced_motion='reduce')
        page.wait_for_timeout(70)
        check('Library respects system reduction but keeps HTML content',page.evaluate("document.querySelector('ft-fall').getStats().motionPresentation==='hide'&&document.querySelector('ft-fall button').getBoundingClientRect().height>0"))
        page.emulate_media(reduced_motion='no-preference')
        page.wait_for_timeout(70)
        page.evaluate("document.querySelector('ft-fall').remove()")
        page.wait_for_timeout(60)
        # Importing just the shared scheduler verifies lifecycle cleanup without loading all renderers.
        check('Removing the falling element clears shared frame jobs',page.evaluate("async()=>{const m=await import('https://toolkit.test/dist/modules/core/scheduler.js');return m.schedulerStats().jobs===0;}"))
        check('No uncaught exceptions across new effects and demo checks',not errors)
        browser.close()
    (OUT/'falling-results.json').write_text(json.dumps({'checks':checks,'distribution':json.loads((ROOT/'dist/manifest.json').read_text())['mode'],'topLevelNavigation':False},indent=2))
    print(f'{len(checks)} falling/status/layout browser checks passed using native Canvas.')


if __name__=='__main__':
    main()
